### ISSUE TITLE
PQC IETF-Hackathon: Composite ML-DSA with EDDSA X.509 structure

### ISSUE THREAD

**Issue by smuellerDD** [bcgit/bc-java#1934]
I am trying to validate my X.509 implementation providing a composite ML-DSA with EDDSA signature against the one in BC as provided in [1] files MLDSA65-Ed25519-SHA512-2.16.840.1.114027.80.8.1.30_ta.der and MLDSA44-Ed25519-SHA512-2.16.840.1.114027.80.8.1.23_ta.der against the implementation in [2].

Albeit the parsing of the ML-DSA and SLH-DSA certificates generated by BC using leancrypto works, the composite ML-DSA X.509 parsing fails.

The exact parsing failure according to my analysis is the data structure of the signature. According to [4] section 6.2, the signature is a SEQUENCE of two entries. Thus, my parser uses the definition [3] for parsing the signature data.

However, my parser fails with the BC X.509 data when starting to parse the signature as the first byte of the signature blob generated by BC is a 0x00 value instead of a 0x30 value indicating a SEQUENCE which my parser is expecting.

May I ask for a pointer where the code is located that generates the composite signature part of the X.509 blob? Or may I ask why the signature block is seemingly no SEQUENCE considering [4] section 6.2?

Thanks a lot

[1] https://github.com/IETF-Hackathon/pqc-certificates/providers/bc/artifacts_certs_r4.zip

[2] https://github.com/smuellerDD/leancrypto

[3] https://github.com/smuellerDD/leancrypto/blob/master/asn1/src/x509_composite_mldsa_signature.asn1

[4] https://www.ietf.org/archive/id/draft-ietf-lamps-pq-composite-sigs-03.html

**Comment by johngray-dev** [comment]
Hi Stephan,

I took a look at the BC certificate that you reference, and I see the Signature section is indeed a BIT STRING (which is required by a X509 Certificate) which is encoded as 03 82 0D 3A 00.   The tag is 3, the length is long form, so 2 bytes, so 0D 3A.  So the Value should be 00 30 82 0D 35 which is supposed to be the SEQUENCE, but the SEQUENCE starts 2 bytes later...   So it does appear there is an extra 00 in there...  Hmm...   I looked at composites I created, and they do the same thing when a BIT STRING is used.  I then looked at a plain vanilla RSA certificate that encodes a BIT STRING for the public key and it does the same thing... The Tag, length items are followed by a 00 when a BIT STRING is used.   This seems to be a consequence of how a DER encoding of a BIT STRING is created.   I think the extra 00 means no extra bytes are padded in the encoded value (because it is already a multiple of 8).   See section 5.4 of this document that describes the encoding:  https://luca.ntop.org/Teaching/Appunti/asn1.html

  If you drop the DER encoding of the Cert into the website https://lapo.it/asn1js/  you will see the initial BIT STRING seems to include the 03 82 03 3A 00 as a separate value, and the next one starts at 30  I took it from this location in our Hackathon repository:  https://github.com/IETF-Hackathon/pqc-certificates/blob/master/providers/bc/artifacts_certs_r4.zip   which is the latest and should conform to -03.   I printed out the ASN.1 structure, and see the SEQUENCE is there.  For example:

SEQUENCE (2 elem)
   BIT STRING (26472 bit) 11111111110001001011001011011011110100100111000101010010011000010111…
   BIT STRING (512 bit) 0000010000000010111101100001010111010101010001000100001100101011010110…

So unless we have all got this wrong for many years,  I think it is probably correct to have the 00 there to indicate no extra padding is used there.
